### PR TITLE
Pytorch trainer for Sempar

### PR DESCRIPTION
**Pull request by rahul1980**
This is a near-identical replication of the DRAGNN trainer and is written completely in python with the exception of the frame-evaluation task.

The top-level trainer is at:
sling/nlp/parser/tools/train_pytorch.py, which depends on a bunch of python files in sling/nlp/parser/trainer and sling/nlp/parser/trainer/flow. While training, it outputs a best checkpoint in the form of a Myelin flow file, which works as-is with the Myelin parser (sling/nlp/parser/tools/parse.cc).

The trainer is not optimized for speed but some optimizations can be handled in future commits.

This PR also fixes a few minor bugs:

Handling of negative indices in pyarray.cc
Handling "True" as true in flow attributes.
Adding a newline at the end of the word-vocab file, otherwise the last word is ignored.

### CODE REVIEW

File: sling/nlp/parser/tools/train_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+      for name, p in sempar.named_parameters():
+        if p.requires_grad:
+          self.averages[name] = p.data.clone()
+
+
+  # Resets the state for a new batch.
+  def _reset(self):
+    self.current_batch_size = 0
+    self.current_batch_num_transitions = 0
+    del self.batch_loss
```
Comment (ringgaard) [inline sling/nlp/parser/tools/train_pytorch.py]: Why do you need a del here? You are overriding this on the next line.

File: sling/nlp/parser/tools/train_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+        torch.nn.utils.clip_grad_norm([p], self.hyperparams.gradient_clip)
+
+
+  # Performs a gradient update.
+  def update(self):
+    if self.current_batch_size > 0:
+      start = time.time()
+      self.batch_loss /= self.current_batch_num_transitions
+
+      # Add the regularization penalty to the batch loss.
```
Comment (ringgaard) [inline sling/nlp/parser/tools/train_pytorch.py]: I did not know that we used L2 regularization. Does it actually help?

File: sling/nlp/parser/trainer/flow/builder.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+from struct import pack
+
+from flow import Variable
+from flow import Function
+from flow import Flow
+
+DT_INT = "int32"
+DT_FLOAT = "float32"
+
+math_kernels = {
```
Comment (ringgaard) [inline sling/nlp/parser/trainer/flow/builder.py]: I think we can delete this, since we are not using this currently.

File: sling/nlp/parser/trainer/flow/builder.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+  def ref(self, instance, var, name=None):
+    r = self.op("Reference", [instance], name)
+    r.producer.add_attr("var", var.name)
+    r.type = var.type
+    r.shape = var.shape
+    return r
+
+
+class Expression:
```
Comment (ringgaard) [inline sling/nlp/parser/trainer/flow/builder.py]: We can also delete the expression class since we are not using it.

File: sling/nlp/parser/trainer/pytorch_modules.py [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+
+  # Returns a string specification of the module.
+  def __repr__(self):
+    s = self.weight.size()
+    return self.__class__.__name__ + "(input_activation=" + str(s[0] - 1) + \
+        ", dim=" + str(s[1]) + ", oov_vector=" + str(s[1])+ ")"
+
+
+# PyTorch replication of the DRAGNN LSTM Cell. This is slightly different from
+# PyTorch's inbuilt LSTM implementation.
```
Comment (ringgaard) [inline sling/nlp/parser/trainer/pytorch_modules.py]: inbuilt -> built-in?

File: sling/nlp/parser/trainer/transition_generator.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+  class Edge:
+    def __init__(self, incoming=None, role=None, value=None):
+      self.incoming = incoming
+      self.role = role
+      self.neighbor = value
+      self.inverse = None
+      self.used = False
+
+
+  # Rough action that will be translated into an Action object later on.
```
Comment (ringgaard) [inline sling/nlp/parser/trainer/transition_generator.py]: Rough action? Are these more violent than normal actions? :) What about SimpleAction?

File: sling/pyapi/pyrecordio.cc [hunk 7]
```diff
@@ -36,6 +36,7 @@ PyMethodDef PyRecordReader::methods[] = {
   {"tell", (PyCFunction) &PyRecordReader::Tell, METH_NOARGS, ""},
   {"seek", (PyCFunction) &PyRecordReader::Seek, METH_O, ""},
   {"rewind", (PyCFunction) &PyRecordReader::Rewind, METH_NOARGS, ""},
+  {"done", (PyCFunction) &PyRecordReader::Done, METH_NOARGS, ""},
```
Comment (ringgaard) [inline sling/pyapi/pyrecordio.cc]: Thanks for  fixing my bugs. Seems like I forgot to add this.

File: sling/nlp/parser/tools/train_pytorch.py (hunk 1 above)
Comment (rahul1980) [reply sling/nlp/parser/tools/train_pytorch.py]: self.batch_loss is the apex of the computation graph. I noticed that if I just reset it (as I do on the next line) and not call del, Python was too lazy to garbage collect the old batch loss (and thus the entire graph). I know that with del, Python might still not return the freed memory to the OS (at least immediately), but empirically I recall seeing much better memory behavior.

File: sling/nlp/parser/tools/train_pytorch.py (hunk 2 above)
Comment (rahul1980) [reply sling/nlp/parser/tools/train_pytorch.py]: We have always used L2=1e-4. This was the default value in DRAGNN's hyperparameter protobuf, which is perhaps why you didn't explicitly see it.  During our initial grid search this was better than no L2 regularization.

File: sling/nlp/parser/trainer/flow/flow.py [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+    input.consumers.append(self)
+
+  def add_output(self, output):
+    """Add output from operation."""
+    self.outputs.append(output)
+    output.producer = self
+
+  def add_attr(self, name, value):
+    """Add operation attribute."""
+    self.attrs[name] = str(value)
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/trainer/flow/flow.py]: Done.

File: README.md [hunk 9]
```diff
@@ -25,7 +25,7 @@ A more detailed description of the SLING parser can be found in this paper:
 * Michael Ringgaard, Rahul Gupta, and Fernando C. N. Pereira. 2017.
   *SLING: A framework for frame semantic parsing*. http://arxiv.org/abs/1710.07032.
 
-</span>
+This is SEMPAR, the first generation of the SLING parser. We have started to work on CASPAR, the second generation of the parser, which can be found [here](https://github.com/google/sling/tree/caspar).
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply README.md]: Indeed. I have reverted this section but kept another which is relevant to CASPAR.

File: sling/myelin/flow.cc [hunk 10]
```diff
@@ -269,7 +269,8 @@ int Attributes::Get(const string &name, int defval) const {
 bool Attributes::Get(const string &name, bool defval) const {
   for (auto &attr : *this) {
     if (attr.name == name) {
-      return attr.value == "1" || attr.value == "T" || attr.value == "true";
+      return attr.value == "1" || attr.value == "T" || attr.value == "true" ||
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/myelin/flow.cc]: Done.

File: sling/nlp/parser/tools/train_pytorch.py [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+
+from functools import partial
+
+sys.path.insert(0, "sling/nlp/parser/trainer")
+
+import train_util as utils
+
+from train_util import mem
+from train_util import now
+from train_util import argparser
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/tools/train_pytorch.py]: Using flags now.

File: sling/nlp/parser/tools/train_pytorch.py [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+# After every 'batch_size' examples, it computes the gradient and applies
+# it, with optional gradient clipping.
+class Trainer:
+  # Training hyperparameters.
+  class Hyperparams:
+    def __init__(self, lr=0.0005, num_examples=1000000, report_every=24000, \
+                 l2_coeff=1e-4, batch_size=8, gradient_clip=1.0, \
+                 optimizer="adam", adam_beta1=0.01, adam_beta2=0.999,
+                 adam_eps=1e-5, moving_avg=True, moving_avg_coeff=0.9999):
+      self.lr = lr
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/tools/train_pytorch.py]: Renamed to alpha.

File: sling/nlp/parser/trainer/flow/builder.py (hunk 3 above)
Comment (rahul1980) [reply sling/nlp/parser/trainer/flow/builder.py]: Done

File: sling/nlp/parser/trainer/flow/builder.py (hunk 4 above)
Comment (rahul1980) [reply sling/nlp/parser/trainer/flow/builder.py]: Done

File: sling/nlp/parser/trainer/pytorch_modules.py (hunk 5 above)
Comment (rahul1980) [reply sling/nlp/parser/trainer/pytorch_modules.py]: Done

File: sling/pyapi/pyrecordio.cc (hunk 7 above)
Comment (rahul1980) [reply sling/pyapi/pyrecordio.cc]: Ack.

**Comment by rahul1980** [comment]
Thanks Michael, I have addressed all your comments. The change to flags.py also made the code slightly more compact.

File: sling/nlp/parser/tools/train_pytorch.py (hunk 1 above)
Comment (ringgaard) [reply sling/nlp/parser/tools/train_pytorch.py]: It might be good to document this in the code since this is not obvious. AFAIK there should normally not be any difference between del and overwriting
